jax/_src/lax/control_flow.py

```diff
@@ -205,9 +206,9 @@ def fori_loop(lower, upper, body_fun, init_val):
   if (isinstance(core.get_aval(lower), ConcreteArray) and
       isinstance(core.get_aval(upper), ConcreteArray)):
     try:
-      lower_ = int(lower)
-      upper_ = int(upper)
-    except TypeError:
+      lower_ = _check_integer("lax.fori_loop", "lower", lower)
+      upper_ = _check_integer("lax.fori_loop", "upper", upper)
+    except TypeError: # change to TracerIntegerConversionError?
```

Comment:
@jakevdp  Could you please give some advice?
Does `fori_loop` not allow `Tracer` with `ConcreteArray` type `aval`?

Reply:
I believe that `int(x)` will return the concrete value if `x` is a tracer with a concrete aval.

Reply:
Again, as far as I can tell this code works correctly and does exactly what we want it to do, so unless you have found a bug somewhere I'd say it's not broken, and there's no need to change it.

Reply:
> Again, as far as I can tell this code works correctly and does exactly what we want it to do, so unless you have found a bug somewhere I'd say it's not broken, and there's no need to change it.

Sorry to be dense about this point, I have little experience in working for a real project. -- BTW, I am curious about how to get a tracer with a concrete aval.

Reply:
It will mainly come in via autodiff. For example:
```python
from jax import grad

def f(x):
  print(x.aval)
  return x

_ = grad(f)(1.0)
# ConcreteArray(1.0, dtype=float32, weak_type=True)
```

Reply:
Thanks for nice example! -- And I find Tracer define appropriate `__int__` method.